packages/xl-ai/src/prosemirror/agent.ts

```diff
@@ -0,0 +1,325 @@
+import { UnreachableCaseError } from "@blocknote/core";
+import { Fragment, Node, Schema, Slice } from "prosemirror-model";
+import { AllSelection, TextSelection, Transaction } from "prosemirror-state";
+import {
+  ReplaceAroundStep,
+  ReplaceStep,
+  Step,
+  Transform,
+} from "prosemirror-transform";
+
+export type AgentStep = {
+  prosemirrorSteps: Step[];
+  selection:
+    | {
+        anchor: number;
+        head: number;
+      }
+    | undefined;
+  type: "select" | "replace" | "insert";
+};
+
+/**
+ * Takes an array of ReplaceSteps and applies them as a human-like agent.
+ * Every step is split into 3 phases:
+ * - select the text to be replaced (1 transaction per ReplaceStep)
+ * - replace the text with the first character of the replacement (if any) (1 transaction per ReplaceStep)
+ * - insert the replacement character by character (strlen-1 transactions per ReplaceStep)
+ */
+export function getStepsAsAgent(doc: Node, pmSchema: Schema, steps: Step[]) {
```

Comment:
This makes more sense to operate on a whole transform, and give a transform back rather than an array of steps:

```ts
export function getStepsAsAgent(trToTransform: Transform) {
  const pmSchema = getPmSchema(trToTransform);
  const { modification } = pmSchema.marks;

  const agentSteps: AgentStep[] = [];

  const tr = new Transform(trToTransform.doc);

  for (const step of trToTransform.steps) {
    ...

return tr;
```

Reply:
I agree on the input, but the current return type does have additional metadata, right? So we can't really return a transform?

Reply:
Yep, a transform wouldn't be enough for that.